HPCC-13051 Redis locking module

jamienoss opened the pull request:
Signed-off-by: James Noss james.noss@lexisnexis.com

jamienoss committed:
HPCC-13051 Redis locking module

Signed-off-by: James Noss <james.noss@lexisnexis.com>

jamienoss commented:
@richardkchapman, @ghalliday please review.

jamienoss commented:
Forward notes from old  PR: #7053 

This is the locking implementation using the hiredis sync API.
This commit also includes some bug fixes for the timeouts (forgot to set upon initial connection).
One or two of the timeouts/expirations had not been converted to secs from mu-s.
Restructured exception message strings, such that any VStringBuffers are only created upon an actual exception.
Added tests to sync code and for locking code.
Added README.
Expire was using DEL instead of EXPIRE preventing PERSIST from working.

The only major outstanding work is whether it is possible to wrap the IF(cond, A, B) code in a macrofunction within the .ecllib library? This way I can remove the EXPORT on the individual functions.

I haven't figured out a way to test the publication method within an actual ECL regression test yet, we can perhaps discuss this today/this week. There are however, some #if code blocks within the C++ to test this manually.

richardkchapman commented:
Is microseconds the appropriate unit to use for timeouts? We usually use milliseconds I think

richardkchapman commented on plugins/redis/lib_redis.ecllib:
```diff
@@ -17,57 +17,82 @@
+END;
+
+EXPORT locking := SERVICE : plugin('redis'), namespace('RedisPlugin')
+  STRING   SetString(  CONST VARSTRING key, CONST STRING value,  CONST VARSTRING options, UNSIGNED database = 0, UNSIGNED4 expire = 0, CONST VARSTRING password = '', UNSIGNED timeout = 1000000) : cpp,action,context,entrypoint='SyncLockRSetStr';
```
I would have called these SetAndPublish... or similar- and the get ones GetOrLock... I would not have separate sync and locking modules.

richardkchapman replied:
Did you decide against including getData/setData here?

jamienoss replied:
I thought you both said no? Does it return NULL or empty etc, I don't know?

jamienoss commented:
Timeout mu-s vs ms? I don't know, I went with freedom, everything in here accepts mu-s. Which would you prefer?

ghalliday commented on testing/regress/ecl/redislockingtest.ecl:
```diff
@@ -0,0 +1,125 @@
+    RETURN IF (LENGTH(value) = 0, myRedis.locking.SetUnicode(key, func(key2)), value);
+END;
+
+CSleep(INTEGER duration) := BEGINC++
```
wouldn't work on non-linux.  We may not care.

ghalliday replied:
More portable is Std.System.Debug.Sleep(duration * 1000);

ghalliday commented on testing/regress/ecl/redislockingtest.ecl:
```diff
@@ -0,0 +1,125 @@
+    myRedis.GetUtf8('utf8'),
+    myRedis.locking.GetUtf8('utf8'),
+    myRedis.FlushDB(),
+    myRedis.locking.SetUtf8('utf8', U'אבגדהוזחטיךכלםמןנסעףפץצקרשת'),
```
Should be using a different string on each different set call.

richardkchapman commented:
Not sure what you mean by "I went with freedom". I would prefer consistency with other timeout settings in our system. Those are generally either seconds or milliseconds.

jamienoss committed:
Changed tests and keys
Changed timeout from mus -> ms

have not removed locking module or renamed those functions, yet.

Signed-off-by: James Noss <james.noss@lexisnexis.com>

jamienoss commented:
@richardkchapman, @ghalliday New commit.

jamienoss committed:
Removed 'locking' module

Renamed locking.Get -> GetOrLock
	locking.Set -> SetAndPublish

Signed-off-by: James Noss <james.noss@lexisnexis.com>

richardkchapman commented on plugins/redis/lib_redis.ecllib:
```diff
@@ -17,57 +17,77 @@
 
 
 EXPORT sync := SERVICE : plugin('redis'), namespace('RedisPlugin')
```
Is it still sensible to call this sync (given there is now no other supported interface to redis)? If sync is something that would be meaningful to a redis user, then I don't mind keeping it...

jamienoss replied:
I wondered this my self and almost changed it to redis, but *if*, for what ever reason we do want to add the async stuff, it would makes sense to keep this as sync now.
Otherwise, no it's not and it's really related to hiredis rather than redis itself.
I'll change to redis.

ghalliday commented on plugins/redis/redis.cpp:
```diff
@@ -0,0 +1,904 @@
+}
+void Connection::connect(ICodeContext * ctx, unsigned __int64 _database, const char * password)
+{
+    struct timeval to = { timeout/1000, timeout%1000 };
```
I think that should be (timeout % 1000) * 1000

ghalliday commented on plugins/redis/README.md:
```diff
@@ -0,0 +1,161 @@
+The **redis-server** package comes with the redis client **redis-cli**. This can be used to send and receive commands to and from the server, invoked by `redis-cli` or, for example,
+`redis-cli -p 6380` to connect to the redis-cache on port 6380 (assuming one has been started).
+
+Perhaps on of the most handy uses of **redis-cli** is the ability to monitor all commands issued to the server via the redis command `MONITOR`. `INFO ALL` is also a useful command
```
typo: on -> one

ghalliday commented on plugins/redis/README.md:
```diff
@@ -0,0 +1,161 @@
+   * `UNSIGNED expire` has a default of **0**, i.e. *forever*.
+
+###The redisServer MODULE
+To avoid the combersom and unnecessary need to constantly pass `options` and `password` with each function call, the module `redisServer` can be imported to effectively 
```
typo: cumbersome

ghalliday commented on plugins/redis/README.md:
```diff
@@ -0,0 +1,161 @@
+requested has not yet been stored, it is desired that only one of such requests obtain the new value and then store it for the others to then also obtain (from the cache).
+This plugin offers a solution to such a problem via the `GetOrLock` and `SetAndPublish` functions within the `redisServer` and `sync` modules of lib_redis.
+This module contains only three function categories - the `SET` and `GET` functions for **STRING**, **UTF8**, and **UNICODE** (i.e. only those that return empty strings)
+and lastley, an auxiliary function `Unlock` used to manually unlock locked keys as it be discussed.
```
typo: lastly

ghalliday commented on plugins/redis/redis.cpp:
```diff
@@ -0,0 +1,904 @@
+
+//The following class is here to ensure destruction of the cachedConnection within the main thread
+//as this is not handled by the thread hook mechanism.
+static class mainThreadCachedConnection
```
trivial: We normally use leading uppercase on class names

jamienoss replied:
Yep true, I'll change.

ghalliday commented on plugins/redis/redis.cpp:
```diff
@@ -0,0 +1,904 @@
+Connection::Connection(ICodeContext * ctx, const char * _options, unsigned __int64 _database, const char * password, unsigned __int64 _timeout)
+  : database(0), timeout(_timeout), port(0)
+{
+    serverIpPortPasswordHash = hashc((const unsigned char*)password, strlen(password), 0);
```
minor: candidate for moving into a function and sharing with the call in isSameConnection

jamienoss replied:
Will do, in fact I think the hash in isSameConnection() is wrong as it is hashing the existing options.

ghalliday commented on plugins/redis/redis.cpp:
```diff
@@ -0,0 +1,904 @@
+    //The following is the dissemination of the two methods authenticate(ctx, password) & selectDB(ctx, _database)
+    //such that they may be pipelined to save an extra round trip to the server and back.
+    if (password && *password)
+      redisAppendCommand(context, "AUTH %b", password, strlen(password));
```
trivial: indented with 2 spaces instead of 4

ghalliday commented on plugins/redis/redis.cpp:
```diff
@@ -0,0 +1,904 @@
+    unsigned hash = hashc((const unsigned char*)options.str(), options.length(), hashc((const unsigned char*)password, strlen(password), 0));
+    return (serverIpPortPasswordHash == hash);
+}
+void * Connection::allocateAndCopy(const char * src, size_t size)
```
trivial: should be a non-member helper function, this is unused.

ghalliday commented on plugins/redis/redis.cpp:
```diff
@@ -0,0 +1,904 @@
+            ctx->logString(msg.str());
+        }
+    }
+    return;
```
trivial: not needed

ghalliday commented on plugins/redis/redis.cpp:
```diff
@@ -0,0 +1,904 @@
+    }
+
+    cachedConnection->Release();
+    cachedConnection = new Connection(ctx, options, _database, password, _timeout);
```
Should set cachedConnection to NULL here.  Otherwise if new Connection fails it will not update cachedConnection, and subsequent calls will access invalid memory.

ghalliday commented on plugins/redis/redis.cpp:
```diff
@@ -0,0 +1,904 @@
+        return;
+    assertConnection();
+    timeout = _timeout;
+    struct timeval to = { timeout/1000, timeout%1000 };
```
same issue with (timeout % 1000) * 1000

ghalliday commented on plugins/redis/redis.cpp:
```diff
@@ -0,0 +1,904 @@
+}
+void Connection::assertOnError(const redisReply * reply, const char * _msg)
+{
+    if (!reply)//assertex(reply)?
```
trivial: comments can be removed

ghalliday commented on plugins/redis/redis.cpp:
```diff
@@ -0,0 +1,904 @@
+}
+void Connection::assertOnCommandErrorWithKey(const redisReply * reply, const char * cmd, const char * key)
+{
+    if (!reply)//assertex(reply)?
```
trivial: comment can be removed

ghalliday commented on plugins/redis/redis.cpp:
```diff
@@ -0,0 +1,904 @@
+    }
+    else if (reply->type == REDIS_REPLY_ERROR)
+    {
+        if (strncmp(reply->str, "NOAUTH", 6) == 0)
```
minor: code repeated 4 times suggests it should be commoned up

ghalliday commented on plugins/redis/redis.cpp:
```diff
@@ -0,0 +1,904 @@
+    if (strncmp(reply->query()->str, REDIS_LOCK_PREFIX, strlen(REDIS_LOCK_PREFIX)) != 0)
+    {
+        //Not locked so return value
+        retVal->set(reply->query()->len, reply->query()->str);
```
minor: if blank it should start again

ghalliday commented on plugins/redis/redis.cpp:
```diff
@@ -0,0 +1,904 @@
+//---------------------------------------------------------------------------------------
+void Connection::encodeChannel(StringBuffer & channel, const char * key) const
+{
+    channel.append(REDIS_LOCK_PREFIX).append("_").append(key).append("_").append(database).append("_").append(ip.str()).append("_").append(port);
```
should the ip be included in the encoded channel?  What if there are multiple balanced redis servers on different ips (question from ignorance)

ghalliday commented on plugins/redis/redis.cpp:
```diff
@@ -0,0 +1,904 @@
+        rtlFail(0, msg.str());
+    }
+
+#if(0)
```
Should this code be deleted?

jamienoss replied:
I'd prefer to keep these for possible future testing if that's ok?

ghalliday commented on plugins/redis/redis.cpp:
```diff
@@ -0,0 +1,904 @@
+    redisAppendCommand(context, cmd.str(), key, strlen(key), value, size);//SET
+    StringBuffer channel;
+    encodeChannel(channel, key);
+    redisAppendCommand(context, "PUBLISH %b %b", channel.str(), channel.length(), value, size);//PUB
```
This means the data is duplicated when sent to thje server.  If it is large that could be inefficient.

ghalliday replied:
Can you assign to a variable, and then use in both commands or something similar using server side scripting?

jamienoss replied:
Yes, good point, I'm on the case now to have this sent to the server once and then set and published from there. From what I read last night it seems fairly trivial to implement.

jamienoss replied:
This will be in 13158 instead of 13155

ghalliday commented on plugins/redis/redis.cpp:
```diff
@@ -0,0 +1,904 @@
+//                           ECL SERVICE ENTRYPOINTS
+//--------------------------------------------------------------------------------
+//-----------------------------------SET------------------------------------------
+ECL_REDIS_API void ECL_REDIS_CALL SyncLockRSetStr(ICodeContext * ctx, size32_t & returnSize, char * & returnValue, const char * key, size32_t valueSize, const char * value, const char * options, unsigned __int64 database, unsigned expire, const char * password, unsigned __int64 timeout)
```
trivial: returnLength instead of returnSize would be consistent with the other functions.

jamienoss replied:
Will be added to 13155 shortly.

ghalliday commented on testing/regress/ecl/redislockingtest.ecl:
```diff
@@ -0,0 +1,122 @@
+//Test compatibiltiy between locking and normal functions.
+SEQUENTIAL(
+    myRedis.FlushDB(),
+    myRedis.SetUtf8('utf8', U'אבגדהוזחטיךכלםמןנסעףפץצקרשת'),
```
All these tests should be using different strings.  (Did I already comment this somewhere else?)

jamienoss replied:
Changed in 13155

ghalliday commented:
@jamienoss a few minor comments/questions post merge.
